In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib
import matplotlib.pyplot as plt
import warnings

In [2]:
#Load data for train and test
train = pd.read_csv('C:/Users/ozhat/Desktop/hw1_fold/data.csv')

In [3]:
train.head()

Id  LotFrontage  LotArea  YearBuilt  YearRemodAdd  TotalBsmtSF  1stFlrSF  \
0   1         65.0     8450       2003          2003          856       856   
1   2         80.0     9600       1976          1976         1262      1262   
2   3         68.0    11250       2001          2002          920       920   
3   4         60.0     9550       1915          1970          756       961   
4   5         84.0    14260       2000          2000         1145      1145   

   2ndFlrSF  GrLivArea  FullBath  ...  KitchenAbvGr  TotRmsAbvGrd  Fireplaces  \
0       854       1710         2  ...             1             8           0   
1         0       1262         2  ...             1             6           1   
2       866       1786         2  ...             1             6           1   
3       756       1717         1  ...             1             7           1   
4      1053       2198         2  ...             1             9           1   

   GarageCars  GarageArea  OpenPorchSF  EnclosedPorch  PoolArea  YrSold  \
0           2         548           61              0         0    2008   
1           2         460            0              0         0    2007   
2           2         608           42              0         0    2008   
3           3         642           35            272         0    2006   
4           3         836           84              0         0    2008   

   SalePrice  
0     208500  
1     181500  
2     223500  
3     140000  
4     250000  

[5 rows x 22 columns]

In [4]:
#shape of train data
train.shape

(1460, 22)

In [5]:
n = train.shape[0] # number of samples
d = train.shape[1] # number of columns/attrinutes
# the last column is y, the lables (the real prices)

## let's check the type of each columns 
## (there are NaNs as well)

In [6]:
#you can also check the data set information using the info() command. 
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 22 columns):
Id               1460 non-null int64
LotFrontage      1201 non-null float64
LotArea          1460 non-null int64
YearBuilt        1460 non-null int64
YearRemodAdd     1460 non-null int64
TotalBsmtSF      1460 non-null int64
1stFlrSF         1460 non-null int64
2ndFlrSF         1460 non-null int64
GrLivArea        1460 non-null int64
FullBath         1460 non-null int64
HalfBath         1460 non-null int64
BedroomAbvGr     1460 non-null int64
KitchenAbvGr     1460 non-null int64
TotRmsAbvGrd     1460 non-null int64
Fireplaces       1460 non-null int64
GarageCars       1460 non-null int64
GarageArea       1460 non-null int64
OpenPorchSF      1460 non-null int64
EnclosedPorch    1460 non-null int64
PoolArea         1460 non-null int64
YrSold           1460 non-null int64
SalePrice        1460 non-null int64
dtypes: float64(1), int64(21)
memory usage: 251.1 KB


# Q1. 

## Create from the dataframe train a design Matrix X and a vector y 

## you can do in in few steps:

In [7]:
df=train.copy() #להוריד שורה ראשונה,עמודה ראשונה ואחרונה, שורות שמופיע בהם נאנים

In [8]:
df = df.dropna()

In [9]:
df.columns

Index(['Id', 'LotFrontage', 'LotArea', 'YearBuilt', 'YearRemodAdd',
       'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'GrLivArea', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd',
       'Fireplaces', 'GarageCars', 'GarageArea', 'OpenPorchSF',
       'EnclosedPorch', 'PoolArea', 'YrSold', 'SalePrice'],
      dtype='object')

In [10]:
df.shape


(1201, 22)

In [11]:
y =df['SalePrice']

In [12]:
df=df.drop(['Id', 'SalePrice'], axis=1) 

In [13]:
y=y.values.reshape(-1,1)

In [14]:
X=df.to_numpy() 

In [15]:
type(X)

numpy.ndarray

In [16]:
# Q2. 

## Create from X and the vector y:

### 1. Estimations of the weights w_hat

### 2. Estimations of y_hat

### 3. mean square error?

### 4. mean absolute error?

### 5. Calculate the matrix that takes w_hat to y_hat

In [17]:
X_t_X=np.dot((np.transpose(X)),X)

In [18]:
from numpy.linalg import inv
inverse=inv(np.matrix(X_t_X))

In [19]:
inverse_Xtran=np.dot(inverse,np.transpose(X))

In [20]:
w_hat=(np.dot(inverse_Xtran,y))

In [21]:
y_hat=np.dot(X,w_hat)

In [22]:
mean_square_error = np.square(np.subtract(y,y_hat)).mean() 

In [23]:
mean_square_error

1655995513.3515604

In [24]:
mean_average_error = np.abs((y_hat-y)).mean()

In [25]:
mean_average_error

25292.980535012386

In [26]:
inv_XtX_Xt = np.dot(inverse,np.transpose(X))

# Q3. 

## Using the weights you got :

### 1. Estimations of the 3 most influential features?

### 2. Estimations of the 3 less influential features?


In [27]:
from heapq import nlargest,nsmallest
val_large= (nlargest(3, np.abs(w_hat)))
influential_features=[]
index = 0
while index < len(val_large):
   influential_features.append(df.columns[list(np.abs(w_hat)).index(val_large[index])])
   index+=1
val_small= (nsmallest(3, np.abs(w_hat)))
neglectible_features=[]
index = 0
while index < len(val_small):
   neglectible_features.append(df.columns[list(np.abs(w_hat)).index(val_small[index])])
   index+=1



In [28]:
influential_features

['KitchenAbvGr', 'GarageCars', 'BedroomAbvGr']

In [29]:
influential_features = ['KitchenAbvGr', 'GarageCars', 'BedroomAbvGr']

In [30]:
neglectible_features

['OpenPorchSF', 'LotArea', 'EnclosedPorch']

In [31]:
neglectible_features = ['OpenPorchSF', 'LotArea', 'EnclosedPorch']

# Q4. 

## Take the same regression problem but now with only 3 features that you found in previous question:

### 1. w_hat?

### 2. y_hat?

### 3. mean sum square error?

### 4. mean sum absolute error?



In [32]:
# calc new design matrix, don't overide the onld matrix X
X3 =df[influential_features]

In [33]:
X3=X3.to_numpy()

In [34]:
inverse_X3tran=np.dot(inv(np.matrix(np.dot((np.transpose(X3)),X3))),np.transpose(X3))

In [35]:
w3_hat =(np.dot(inverse_X3tran,y))

In [36]:
w3_hat

matrix([[-9750.34005156],
        [73688.19945664],
        [20536.70180197]])

In [37]:
y3_hat = np.dot(X3,w3_hat)  

In [38]:
y3_hat

matrix([[199236.16426762],
        [199236.16426762],
        [199236.16426762],
        ...,
        [146084.66661295],
        [105011.26300901],
        [125547.96481098]])

In [39]:
mean_square_error3 = np.square(np.subtract(y,y3_hat)).mean()

In [40]:
mean_square_error3

4000170588.9776464

In [41]:
# mean_average_error3 = np.sum(np.absolute(y3_hat-y))
mean_average_error3 = np.abs((y3_hat-y)).mean()

In [42]:
mean_average_error3

43991.34675178245

# Q5. 

## Repeat Q2 but now with only half of the samples 

### (sampled randomly for the rows of the original X and y which had 1460 rows)

### 1. Estimations of the weights w_hat

### 2. Estimations of y_hat

### 3. mean square error?

### 4. mean absolute error?

### 5. Calculate the matrix that takes w to x_hat

In [43]:
X5=train[train['Id']<=730]

In [44]:
X5=X5.dropna()# X5=X5[X5['LotFrontage']>=0] .dropna()

In [45]:
X5.shape

(606, 22)

In [46]:
y5 =X5['SalePrice']

In [47]:
X5=X5.drop(['Id', 'SalePrice'], axis=1)


In [48]:
y5=y5.values.reshape(-1,1)

In [49]:
X_t_X5=np.dot((np.transpose(X5)),X5)

In [50]:
inverse=inv(np.matrix(X_t_X5))

In [51]:
inverse_Xtran=np.dot(inverse,np.transpose(X5))

In [52]:
w5_hat=(np.dot(inverse_Xtran,y5))

In [53]:
y5_hat=np.dot(X5,w5_hat)

In [54]:
mean_square_error5 = np.square(np.subtract(y5,y5_hat)).mean()

In [55]:
mean_square_error5

1362620677.5848866

In [56]:
mean_average_error5 =np.abs((y5_hat-y5)).mean()

In [57]:
mean_average_error5

24130.66030308007

In [58]:
inv_XtX_Xt = np.dot(inverse,np.transpose(X5))

In [59]:
# Q6. 

## Compare the results between all 3 models that you built

In [60]:
models_list=['m1-original','m2-3 columns','m3-half samples']

In [61]:
arr_mse=[mean_square_error,mean_square_error3,mean_square_error5]

In [62]:
arr_mae=[mean_average_error,mean_average_error3,mean_average_error5]

In [63]:
index_sorted_mse=np.argsort(arr_mse)
index_smallest_mse=index_sorted_mse[0]
index_medium_mse=index_sorted_mse[1]
index_biggest_mse=index_sorted_mse[2]
print("The smallest MSE:",arr_mse[index_smallest_mse]," of ", models_list[index_smallest_mse] )
print("The medium MSE:",arr_mse[index_medium_mse]," of ", models_list[index_medium_mse] )
print("The biggest MSE:",arr_mse[index_biggest_mse]," of ", models_list[index_biggest_mse] )


The smallest MSE: 1362620677.5848866  of  m3-half samples
The medium MSE: 1655995513.3515604  of  m1-original
The biggest MSE: 4000170588.9776464  of  m2-3 columns


In [64]:
index_sorted_mae=np.argsort(arr_mae)
index_smallest_mae=index_sorted_mae[0]
index_medium_mae=index_sorted_mae[1]
index_biggest_mae=index_sorted_mae[2]
print("The smallest MAE:",arr_mae[index_smallest_mae]," of ", models_list[index_smallest_mae] )
print("The medium MAE:",arr_mae[index_medium_mae]," of ", models_list[index_medium_mae] )
print("The biggest MAE:",arr_mae[index_biggest_mae]," of ", models_list[index_biggest_mae] )

The smallest MAE: 24130.66030308007  of  m3-half samples
The medium MAE: 25292.980535012386  of  m1-original
The biggest MAE: 43991.34675178245  of  m2-3 columns
